### PR TITLE
Center the coordinates to pixels for rasterio backend

### PR DESCRIPTION
**Pull request by gbrener**
Rasterio uses edge-based coordinates, which is a different convention from how xarray treats coordinates (based on description here: http://xarray.pydata.org/en/stable/plotting.html#coordinates). This PR centers them, offsetting by half of the resolution.

 - [X] Tests added / passed
 - [X] Passes ``git diff upstream/master | flake8 --diff``
 ~- [ ] Closes #xxxx~
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

CCing @fmaussion since he may be interested.

### CODE REVIEW

**Comment by jbednar** [comment]
Looks good to me, and very important...

**Comment by jhamman** [comment]
Thanks for the PR. I'll let @fmaussion comment on the substance here but given the confusion and the change in behavior, let's add a note in `whats-new.rst` and in the `open_rasterio()` doc string.

**Comment by gbrener** [comment]
Thanks for the feedback @jhamman - just added the documentation.

File: doc/whats-new.rst [hunk 1]
```diff
@@ -21,6 +21,10 @@ v0.9.7 (unreleased)
 Enhancements
 ~~~~~~~~~~~~
 
+- :py:func:`~xarray.open_rasterio` method now shifts the rasterio
+  coordinates so that they are centered in each pixel.
+  By `Greg Brener <https://github.com/gbrener>`_.
```
Comment (fmaussion) [inline doc/whats-new.rst]: This change should rather appear in "Bug fixes" I think

File: xarray/backends/rasterio_.py [hunk 2]
```diff
@@ -87,7 +87,10 @@ def open_rasterio(filename, chunks=None, cache=None, lock=None):
 
     This should work with any file that rasterio can open (most often:
     geoTIFF). The x and y coordinates are generated automatically from the
-    file's geoinformation.
+    file's geoinformation, shifted to the center of each pixel (see
+    `"PixelIsArea" Raster Space
+    <http://web.archive.org/web/20160326194152/http://remotesensing.org/geotiff/spec/geotiff2.5.html#2.5.2>`_
```
Comment (fmaussion) [inline xarray/backends/rasterio_.py]: Thanks for looking into this. Isn't there a better ref than on the web archive?

File: xarray/backends/rasterio_.py [hunk 3]
```diff
@@ -132,8 +135,12 @@ def open_rasterio(filename, chunks=None, cache=None, lock=None):
     dx, dy = riods.res[0], -riods.res[1]
     x0 = riods.bounds.right if dx < 0 else riods.bounds.left
     y0 = riods.bounds.top if dy < 0 else riods.bounds.bottom
-    coords['y'] = np.linspace(start=y0, num=ny, stop=(y0 + (ny - 1) * dy))
-    coords['x'] = np.linspace(start=x0, num=nx, stop=(x0 + (nx - 1) * dx))
+    coords['y'] = np.linspace(start=y0 + dy/2,
+                              num=ny,
```
Comment (fmaussion) [inline xarray/backends/rasterio_.py]: Nit: unnecessary linebreak

File: xarray/backends/rasterio_.py [hunk 4]
```diff
@@ -132,8 +135,12 @@ def open_rasterio(filename, chunks=None, cache=None, lock=None):
...
+                              stop=(y0 + (ny - 1) * dy) + dy/2)
+    coords['x'] = np.linspace(start=x0 + dx/2,
+                              num=nx,
```
Comment (fmaussion) [inline xarray/backends/rasterio_.py]: Nit: unnecessary linebreak

**Comment by fmaussion** [comment]
Definitely a bad bug which could justify a minor release soon :-(

File: xarray/backends/rasterio_.py (hunk 2 above)
Comment (gbrener) [inline xarray/backends/rasterio_.py]: Happy to help @fmaussion . The web archive link is used by the GeoTIFF website (first link here: http://trac.osgeo.org/geotiff). A quick Google search didn't yield anything better.

**Comment by gbrener** [comment]
Thanks for the feedback @fmaussion . I addressed your suggestions and made some small cleanups. Please let me know if there's anything else.

File: xarray/backends/rasterio_.py (hunk 3 above)
Comment (gbrener) [inline xarray/backends/rasterio_.py]: This was just to satisfy `flake8`'s line-length requirement. The three lines were for consistency, but I changed it to be two lines.

File: xarray/backends/rasterio_.py (hunk 4 above)
Comment (gbrener) [inline xarray/backends/rasterio_.py]: This was just to satisfy `flake8`'s line-length requirement. The three lines were for consistency, but I changed it (in my recent commit) to be two lines.

**Comment by shoyer** [comment]
I'm not sure we really have a cell centering convention for xarray, except for plotting routines, where we chose the cell-centered convention from CF-conventions. But this does seem like a marginal improvement, so +1 from me.

In the future we may want to switch to describing bounds fully using `IntervalIndex` from the latest release of pandas. But the API there is still being sorted out along with some bug fixes (see https://github.com/pandas-dev/pandas/pull/16386), so it's probably better to stick with this for now.

**Comment by PeterDSteinberg** [comment]
+1 from me as well.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**